### ISSUE TITLE
Add in-repo project: ToC-generating program for this project

### ISSUE THREAD

**Issue by JordanMartinez** [JordanMartinez/purescript-jordans-reference#222]
#43 has the idea of a single file that acts like a clickable ToC that will open up the corresponding file, making navigation much easier overall.

I'd like to target this concept as the next in-repo project to work on. I think the libraries it should cover are:
- parsing
- file I/O
- conduit
- yargs

Thus, I will write a program that generates outputs/updates a file with an up-to-date Table of Contents of this entire repo that, when an entry is clicked, will open the corresponding file in GitHub on the `latestRelease` branch.

While it will produce a GitHub-friendly version by default, it should be written with the option of generating a ToC using links for your own local computer.

I also want it to include all section headers, so that one can literally see the overall structure of this repo in one page. If I get there, I'll add "sections" in the Syntax folder so that those can be viewed, too.

I'm not sure what the final look of the ToC should be since there will be many intermediate folder names involved and I'd like to show the structure of things without it getting too long.

**Comment by JordanMartinez** [comment]
Purpose
- build a program that can parse this project's files and auto-create a ToC file

Principles
- Can be expanded to work with other source-code-hosting sites (e.g. BitBucket) or other URL formats for section headers / line numbers

Outcome 1
- User runs a program with a whitelist filter as a way to determine which files to include...
    - Since we're only concerned about markdown and PS files, we can whitelist `*.purs` or `*.md` files so that everything else gets excluded by default (hidden files/folders (starts with "."), dependency manager files ("psc-package.json")).
- ...and a file location/name....
    - this allows us to change the file name or location if/when we should desire.
- The program outputs or overwrites that file with a correct ToC that, when a URL is clicked, will open the corresponding page for this repo on GitHub...
    - Why GitHub and not something else? Because this repo is hosted on GitHub
    - Thus, the links that are generated will need to work with GitHub's link specifications:
        - For markdown files, this will use the regular markdown format (i.e. `file-name.md#section-header`).
        - For PS files (e.g. Syntax folder), it will highlight the corresponding section's header by referring to its line in that file (i.e. `file-name.purs#L4`).
- ...using a tag.
    - The ToC should only be updated when a new release is made. Thus, the `development` branch's ToC file will always be outdated. Likewise, the `latestRelease` branch will be correct only for one release. When a new major release is made, previous releases would no longer refer to the correct pages if files' names or locations had changed, leading to broken links. Thus, the links should refer to a tag on GitHub rather than a branch name or a commit hash, so that each release always refers to itself as the version to use for finding files.

The format of the ToC could appear in two different ways:
- As a expandable outline using nested `<details><summary> ... </details>` HTML tags
    - This is easier to scroll through, but comes at the cost of collapsing the summary tags every time the page reloads/refreshes
- As a single outline using nested bullet points
    - This will make the page longer, so it might help to have a ToC for each top-level folder's section.

The format of the ToC should look like this:

<hr>

## Table of Contents

- [link to each top-level folder's section in this file]()

## Top-Level Folder

- [ReadMe]() - if it exists
- [01-file]()
- [02-file]()
- 03-folder - no hyperlink
    - [ReadMe]() - if it exists
    - [01-file]()
    - 02-folder - no hyperlink
        - [ReadMe]() - if it exists
        - [01-file]()
- 04-folder - no hyperlink
    - [ReadMe]() - if it exists
    - 01-folder - no hyperlink
        - [ReadMe]() - if it exists
        - [01-file]()

<hr>

An example using current project setup

<hr>

## Table of Contents

- [Getting Started]()
- [Build Tools]()
- [Syntax]()
- ...

## Getting Started

- [ReadMe]()
- [01-Install-Guide.md]()
- [02-The-REPL.md]()
- [03-Other-Important-Info.md]()

## Build Tools

- [ReadMe]()
- 01-Tool-Comparisons
    - [01-Dependency-Managers.md]()
    - [02-Build-Tools.md]()
- 02-CLI-Options
    - [ReadMe.md]()
- [03-New-Projects-From-Start-To-Finish.md]()
- ...

...

<hr>

Outcome 2

- Learner has a streamlined process for understanding how each component/library builds up into the final product

Brainwrite
- Make a new release that includes the ToC file and the program that does it
- Update release checklist to insure a new release has an updated ToC before being released
- Clean up explanations to appear more streamlined
- Finish fixing any bugs or other things that should have been included in the original specification that were not
- Review code for any changes that need to be made to the original specification and implement them
- Finish explaining libraries and other things needed to understand how code works and why
- Finish writing initial program
- Study libraries needed to make program work
- Specify how a parser will identify a header in some file for both Markdown and PureScript files
- Design overall flowchart of the program itself

**Comment by JordanMartinez** [comment]
Update on this project:

First, I've learned how to use yargs, node-path, node-fs-aff, and `string-parsers`. The are still a few things with `string-parsers` that I don't understand. Since I didn't need to use them AFAIK, I'm not going to cover them further. Also, I did not figure out how to produce great parsing errors, so sometimes it was difficult to track down a bug/issue in the parser code itself.

Second, I realized that the nested nature of these headers can get pretty complicated to parse when one includes multi-line comments. For example,
```purescript
------------
-- These are not an issue...

-- ## 1 - This is a header

-- ### 1.1 - This is a header

-- ### 1.2 - This is a header

------------
-- These possibilities do present an issue

-- ## 2 - This is a header

{-

### 2.1 - This is a header

-}

{- ### 2.2 - This is a header -}

{-
#### 2.2.1 - This is a header

#### 2.2.2 - This is a header

-}

-- #### 2.2.3 - This is a header
```

Third, there are situations where I might include '#' in a comment as a reference to the 'applyFlipped' function name: `arg # function`. I believe that occurs sometimes in a multi-line comment, so it makes it much harder to distinguish a header prefix from a regular function.

Thus, to keep things simple (though this could be improved in the future), headers in PS files will look like this:
```purescript
-- ####... Header Text on regular single-line comment

{-
-- ####... Header Text inside a block-level / multi-line comment: make it look like a single-line comment
-}

{- -- ###... this is not a header -}
```

**Comment by JordanMartinez** [comment]
Another update on this:

- I can generate a ToC that includes headers inside of files but only for PS files. I'm still working on integrating the other two types of files I want (JS and markdown)
- the output isn't sorted correctly (where ReadMe files come before the numbered files)
- most of this repo doesn't have headers inside of its PS files, so I'll need to add those as well.
- I'm currently using hard-coded values to generate the ToC. I need to integrate my program with Yargs to make it more flexible. (For example, I could include a ToC file for each folder in this project.

**Comment by JordanMartinez** [comment]
## Update Summary

The program works!

Things I've fixed since the last update:
- ReadMe files now come before the numbered ones.
- both headers in PS and Markdown files are rendered now due to changing the type I provide to the renderer (a function, not a data type).

Things I still have to do:
- I have not yet integrated my program with Yargs, but this will likely be done by the end of next week
- the `benchmark` folder appears before the `src` and `test` folders when it should appear afterwards if one wants to read this in a natural ordering. I might work on this by providing better sorting options before the content gets rendered or I might not because that only improves the result by a small factor.
- I haven't written the code that makes this work in the Free/Run monad application structure style. It is only encoded in the ReaderT design pattern thus far.
- I have not written the beginner-friendly materials that would explain the fundamental concepts I use and the functions/algorithms I compose to make this program work.
- Write tests that insure all links work correctly, so I can run it via CI when making a new release.

Other issues I found:
- Sometimes, I use an `h1` for something and an `h3` follows it rather than an `h2`. The current version of the program will indent such items four spaces too much and the result will be understood as code, not a list item in a bulleted list.
- Due to how long the output is, I'm not sure that adding headers for each PS file will be worth it. I'm also wondering whether I should use a different format to make it more manageable.

## Current Output

You can see the output in the temporary table-of-contents.md file (link below). Be aware of these things when using that link:
- It's quite long!
- All links point to the `development` branch, not the `latestRelease` branch. 
- I have not verified that all links work

[The temporary `table-of-contents.md` file in the development branch](https://github.com/JordanMartinez/purescript-jordans-reference/blob/development/table-of-contents.md).

**Comment by JordanMartinez** [comment]
Latest update:
- Yargs has been integrated into program
- I've written some beginner-friendly materials for Yargs, Node-FS, purescript-tree, purescript-string-parsers, but these materials need to be edited and refined
- The code now appears in the development branch: `Projects/src/Table of Contents/ToC/`

Things left to do in this issue:
- update code, so that it also tests whether generated links are correct
    at the same time it is creating them.
- update design thought process file to match actual code
- rewrite code using Run application structure
- improve error handling
    - if the header parser fails to parse a header, program should crash or recover from error or something...
    - if the file's hyperlink doesn't return a 200 HTTP response, the program should crash or recover from error or something...
- write tests for various things

**Comment by JordanMartinez** [comment]
Latest update on this. I've refactored the code a bit in an attempt to verify that the file's URLs are valid. I encountered a 'Maximum Call Stack Size Exceeded' error, and it took some time to realize that the problem was caused by referring to the wrong variable.

However, there was some good news from this:
- I learned how to use `tailRec` and understand why/when one should use it
- I learned the value of naming the `go` function as a different name when its used throughout a project as it makes tracking down any stack-size errors easier to find.

Overall, I don't think I like the program logic of my current not-yet-pushed design. On one hand, I'm trying to make it fast. On the other hand, that desire might make the code much harder for a new learner to understand.

**Comment by JordanMartinez** [comment]
Latest update on this.

In my initial design of this program, I walked the file tree once to parse the content and I walked the resulting structure a second time to render it. After having a better understanding of the design space, I saw that that was largely wasteful in CPU cycles. In the current implementation (now pushed to the `development` branch), I walk the file system once, parse and render files, and render directories as I return to its parent directory. The previous approach was more like OO in FP syntax whereas this second approach is more like the FP declarative programming style via `traverse`. (I guess that's why they say that `Traversable` is the answer to everything...)
On top of that, I made it run in parallel thanks to `Parallel` and `Aff`. (Actually, running it in parallel  means my logger is now incorrect.... Ha! Whoops!)
I am also checking the file's URLs for a 200 HTTP code before I render them, so that if it fails, then I just render a plain text entry for that item in the resulting file. If it succeeds, I'll render it as a hyperlink using that file.

The code could still be cleaned up in a few ways:
- I'd like to upgrade the parser to nest headers automatically in the parser rather than use my current parse-all-then-nest approach. I started doing this but it got complicated quickly and I realized it would not be readily understandable to new learners. So, I'm still using the current approach.
- Yargs is no longer the CLI library I want to use now that [opt-parse](https://github.com/f-o-a-m/purescript-optparse) has been ported to PureScript
- Due to how this code is now structured, the Core type files could be merged into one file, the `API` folder might as well be removed and make the `AppM` file take its place.

Before making a new release, Travis CI should now run this program and fail if a single link is considered invalid

On top of that, there's a few other issues relating to new people reading this for the first time:
- the Design Thought Process file now needs to be re-updated to explain why the code was written that way
- I need to explain more pre-reqs to this program: Parallel and Traversable (#178)
